### PR TITLE
CELEBORN-1689: CLI Support for Ratis APIs

### PR DESCRIPTION
**Pull request by akpatnam25**
### What changes were proposed in this pull request?
Add CLI support for Ratis REST APIs that were recently added in https://issues.apache.org/jira/browse/CELEBORN-1628
Leaving out local raft meta conf APIs for no since those are not as straightforward.  Will update those in another PR. 

### Why are the changes needed?
better ops since with REST API and CLI there is no need to log onto production hosts to manage ratis operations. 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
added unit tests

### CODE REVIEW

**Comment by SteNicholas** [comment]
@akpatnam25, why not directly use ratis shell of ratis? The REST API of ratis is same as the ratis shell.

**Comment by akpatnam25** [comment]
> @akpatnam25, why not directly use ratis shell of ratis? The REST API of ratis is same as the ratis shell.

To use ratis shell, you need to be on the host itself, and also have access to the distribution to run it. I think in most production deployments, it is not convenient (and sometimes not even possible) to access the host and run it within the host. I think from operational point of view, it is better to use the REST API  via the CLI since it can be accessed from anywhere outside host. Let me know what you think :) @SteNicholas

**Comment by SteNicholas** [comment]
@akpatnam25, we would use ratis-shell in master node for our production pratice. Meanwhile, CLI does not depend on the REST API, otherwise why not directly invoke the REST API?

**Comment by akpatnam25** [comment]
> @akpatnam25, we would use ratis-shell in master node for our production pratice. Meanwhile, CLI does not depend on the REST API, otherwise why not directly invoke the REST API?

@SteNicholas  CLI does depend on REST Api. It calls the REST api. Not all users would have access to their production host, and would only have access to REST api.

**Comment by akpatnam25** [comment]
cc @FMX

**Comment by akpatnam25** [comment]
cc @turboFei

**Comment by akpatnam25** [comment]
@SteNicholas  I think for the ratis API it makes sense to have cli support for them and we can deprecate the ratis-shell if that is ok with the community later on. We don’t need to support for all rest api’s but I think this one is especially important given the only other way to do it is to log onto a production host. We do not need cli support for every single rest api, just the main ones that help with operational management. Let me know :) thanks.

File: .idea/vcs.xml [hunk 1]
```diff
@@ -1,40 +1,36 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<!--
```
Comment (FMX) [inline .idea/vcs.xml:2]: These changes are not needed.

File: cli/src/main/scala/org/apache/celeborn/cli/master/MasterOptions.scala [hunk 2]
```diff
@@ -120,4 +120,46 @@ final class MasterOptions {
     names = Array("--delete-apps"),
     description = Array("Delete resource of an application."))
   private[master] var deleteApps: Boolean = _
+
```
Comment (FMX) [inline cli/src/main/scala/org/apache/celeborn/cli/master/MasterOptions.scala:123]: There are two more commands to implement.
- show raft group info 
- print local raft meta conf

File: cli/src/test/scala/org/apache/celeborn/cli/TestCelebornCliCommands.scala [hunk 3]
```diff
@@ -273,6 +274,90 @@ class TestCelebornCliCommands extends CelebornFunSuite with MiniClusterFeature {
...
+  test("master --transfer-ratis-leader") {
+    val args = prepareMasterArgs() ++ Array(
+      "--transfer-ratis-leader",
+      "--peerAddress",
+      s"${master.masterArgs.host}:9872")
+    // This test just makes sure the right API response is received since HA is not enabled for unit tests.
+    captureOutputAndValidateResponse(
+      args,
+      "",
+      "org.apache.celeborn.service.deploy.master.clustermeta.SingleMasterMetaManager" +
```
Comment (FMX) [inline cli/src/test/scala/org/apache/celeborn/cli/TestCelebornCliCommands.scala:286]: This test is invlid. You can setup a minu cluster to test ratis cluster.

Comment (FMX) [reply cli/src/test/scala/org/apache/celeborn/cli/TestCelebornCliCommands.scala:286]: You can refer to `MiniClusterFeature`.

File: cli/src/main/scala/org/apache/celeborn/cli/master/MasterOptions.scala [hunk 4]
```diff
@@ -120,4 +120,46 @@ final class MasterOptions {
...
+  private[master] var removeRatisPeers: Boolean = _
+
+  @Option(
+    names = Array("--set-ratis-peers-priorities"),
+    description = Array(
+      "Set the priority of the peers in the raft group. Specify priorities via --priorities."))
+  private[master] var setRatisPeersPriorities: Boolean = _
+
+  @Option(
+    names = Array("--create-snapshot"),
```
Comment (turboFei) [inline cli/src/main/scala/org/apache/celeborn/cli/master/MasterOptions.scala:162]: how about using similar command names?

image

```
--election-transfer
--election-step-down
--election-pause
--election-resume

--peer-add
--peer-remove
--peer-set-priority

--snapshot-create

--local-raft-conf
```

**Comment by akpatnam25** [comment]
spoke more with @SteNicholas - given that ratis shell is the source of truth, we  can just keep it like that. We would not need CLI support for these. Closing out this PR. thanks for the feedback to everyone

**Pull request closed without merging**